src/Statistics/Coverage.cpp

```diff
@@ -0,0 +1,124 @@
+/* Copyright 2021 Coderrect Inc. All Rights Reserved.
+Licensed under the GNU Affero General Public License, version 3 or later (“AGPL”), as published by the Free Software
+Foundation. You may not use this file except in compliance with the License. You may obtain a copy of the License at
+https://www.gnu.org/licenses/agpl-3.0.en.html
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an “AS IS” BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "Coverage.h"
+
+#include <llvm/Support/FormatVariadic.h>
+
+#include "Trace/ProgramTrace.h"
+
+using namespace race;
+
+namespace {
+
+// return true if the function is “external”
+bool isExternal(const llvm::Function *fn) { return fn->isDeclaration() || fn->isIntrinsic(); }
+
+// a function signature = return val type (param type(s)) function name
+// cannot find a better way to compute signature of fn without transferring between std::string and llvm::StringRef
+std::string getSignature(const Function *fn) {
+  llvm::FunctionType *typ = fn->getFunctionType();
+  std::string s;
+  llvm::raw_string_ostream os(s);
+  typ->print(os);
+  os << " " << fn->getName();
+  os.str();
+
+  return s;
+}
+
+void recordFn(std::map<std::string, const llvm::Function *> &map, const llvm::Function *fn) {
+  if (fn == nullptr || isExternal(fn)) return;
+  std::string sig = getSignature(fn);
+  auto exist = map.find(sig);
+  if (exist == map.end()) {
+    map.insert(std::make_pair(sig, fn));
+  }
+}
+
+}  // namespace
+
+Coverage::Coverage(const ProgramTrace &program) : program(program), module(program.getModule()) {
+  summarize();
+  computeFnCoverage();
+}
+
+void Coverage::summarize() {
+  if (!data.analyzed.empty() || !data.total.empty()) return;  // already computed
+
+  // collect fns in module
+  for (auto const &func : module.getFunctionList()) {
+    auto name = func.getName();
+    auto fn = module.getFunction(name);
+    recordFn(data.total, fn);
+  }
+
+  // collect fns in program
+  for (auto const &thread : program.getThreads()) {
+    auto _1stEvent = thread->getEvents().front().get();
+    recordFn(data.analyzed, _1stEvent->getFunction());
+
+    for (auto const &event : thread->getEvents()) {
+      switch (event->type) {
+        case Event::Type::Call: {
+          auto call = llvm::cast<EnterCallEvent>(event.get());
+          auto fn = call->getCalledFunction();
+          recordFn(data.analyzed, fn);
+          break;
+        }
+        case Event::Type::Fork: {
+          auto fork = llvm::cast<ForkEvent>(event.get());
+          auto call = llvm::cast<llvm::CallBase>(fork->getInst());
+          if (OpenMPModel::isFork(call)) {
+            data.numOpenMPRegions++;
+          }
```

Comment:
This check can be done as:

```suggestion
          if (event->getIRType() == IR::Type::OpenMPFork) {
            data.numOpenMPRegions++;
          }
```

Reply:
Yes, I'll update this in other branches? or restore the branch?